### PR TITLE
Add autosaving to the editor

### LINKED ISSUE #686: warn when leaving edit page with unsaved change
**Is your feature request related to a problem? Please describe.**
see [this toot](https://makerdon.org/@Sylphox/103041707542858913)

**Describe the solution you'd like**
either an auto-saving feature, but it would probably require some form of versioning to not publish half a modification when editing an already published post, or simply a warning popup about losing changes when leaving the page (can be done easily in js, see [this anwser on stack overflow](https://stackoverflow.com/questions/7317273/warn-user-before-leaving-web-page-with-unsaved-changes))

~~**Describe alternatives you've considered**~~

~~**Additional context**~~

### PR DESCRIPTION
**Pull request by epsilon-phase**
This should fix #686 Confirmation prior to navigating away is easy enough to add, but it'd be good to know if it's wanted first since this seems like it might serve the same purpose

### CODE REVIEW

File: plume-front/src/editor.rs [hunk 1]
```diff
@@ -100,6 +216,11 @@ fn filter_paste(elt: &HtmlElement) {
 }
 
 pub fn init() -> Result<(), EditorError> {
+    if let Some(ed) = document().get_element_by_id("plume-fallback-editor") {
+        window().set_timeout(autosave, 15000);
```
Comment (fdb-hiroshima) [inline plume-front/src/editor.rs]: I guess set_interval does not exist? :confused:

File: plume-front/src/editor.rs [hunk 2]
```diff
@@ -62,7 +64,121 @@ impl From<stdweb::private::ConversionError> for EditorError {
         EditorError::TypeError
     }
 }
-
+#[derive(Serialize, Deserialize)]
+struct AutosaveInformation {
+    contents: String,
+    title: String,
+    last_saved: f64,
+}
```
Comment (fdb-hiroshima) [inline plume-front/src/editor.rs]: while content is probably the most annoying to loose, if we do a local save, we should as well save everything (including subtitle, tags...)

File: plume-front/src/editor.rs (hunk 1 above)
Comment (epsilon-phase) [reply plume-front/src/editor.rs]: It doesn't look like it, oddly.

File: plume-front/src/editor.rs (hunk 2 above)
Comment (epsilon-phase) [reply plume-front/src/editor.rs]: True enough, easy enough to add too

File: plume-front/src/editor.rs (hunk 1 above)
Comment (epsilon-phase) [reply plume-front/src/editor.rs]: Okay, works fine in the `js!`

File: plume-front/src/editor.rs [hunk 3]
```diff
@@ -62,7 +69,183 @@ impl From<stdweb::private::ConversionError> for EditorError {
...
+fn get_cover() -> String {
+    let cover: SelectElement =
+        SelectElement::try_from(document().get_element_by_id("cover").unwrap())
+            .ok()
+            .unwrap();
+    console!(log, "Got cover");
+    cover.raw_value()
+}
+fn autosave() {
+    let info: AutosaveInformation = AutosaveInformation {
```
Comment (AnaGelez) [inline plume-front/src/editor.rs]: Is the type hint really needed?

File: plume-front/src/editor.rs [hunk 4]
```diff
@@ -62,7 +69,183 @@ impl From<stdweb::private::ConversionError> for EditorError {
...
+        .ok()
+        .unwrap();
+    tags.raw_value()
+}
+fn get_license() -> String {
+    let license: InputElement =
+        InputElement::try_from(document().get_element_by_id("license").unwrap())
+            .ok()
+            .unwrap();
+    console!(log, "Got license");
```
Comment (AnaGelez) [inline plume-front/src/editor.rs]: Maybe this line should be removed ?

File: plume-front/src/editor.rs [hunk 5]
```diff
@@ -62,7 +69,183 @@ impl From<stdweb::private::ConversionError> for EditorError {
...
+    license.raw_value()
+}
+fn get_cover() -> String {
+    let cover: SelectElement =
+        SelectElement::try_from(document().get_element_by_id("cover").unwrap())
+            .ok()
+            .unwrap();
+    console!(log, "Got cover");
```
Comment (AnaGelez) [inline plume-front/src/editor.rs]: This one too

File: plume-front/src/editor.rs [hunk 6]
```diff
@@ -62,7 +69,183 @@ impl From<stdweb::private::ConversionError> for EditorError {
...
+    if let Some(autosave_str) = window().local_storage().get(&get_autosave_id()) {
+        let autosave_info: AutosaveInformation = serde_json::from_str(&autosave_str).ok().unwrap();
+        let message = format!(
+            "{} {}?",
+            i18n!(
+                CATALOG,
+                "Do you want to load the local autosave last edited at"
+            ),
+            Date::from_time(autosave_info.last_saved).to_date_string()
+        );
```
Comment (AnaGelez) [inline plume-front/src/editor.rs]: You should use `i18n!` for formatting, to allow the date to be placed somewhere else in the sentence if translators need to do so.

```rust
i18n!(
    CATALOG,
    "Do you want to load the local autosave last edited at {}";
    Date::from_time(autosave_info.last_saved).to_date_string()
)
```

File: plume-front/src/editor.rs [hunk 7]
```diff
@@ -62,7 +69,183 @@ impl From<stdweb::private::ConversionError> for EditorError {
...
+        }
+    }
+}
+fn clear_autosave() {
+    window().local_storage().remove(&get_autosave_id());
+    console!(log, &format!("Saved to {}", &get_autosave_id()));
+}
+static mut AUTOSAVE_TIMEOUT: Option<TimeoutHandle> = None;
 
+unsafe fn autosave_debounce(interval: u32) {
```
Comment (AnaGelez) [inline plume-front/src/editor.rs]: Why does it need to be `unsafe`?

File: plume-front/src/editor.rs (hunk 3 above)
Comment (epsilon-phase) [reply plume-front/src/editor.rs]: Oh, probably not

File: plume-front/src/editor.rs (hunk 7 above)
Comment (epsilon-phase) [reply plume-front/src/editor.rs]: Oh, it should probably just be unsafe on the lines that refer to the `static mut`

Comment (AnaGelez) [reply plume-front/src/editor.rs]: No, I don't know much about unsafe, so maybe you are right and whole function should be. But also, I feel like there is a better way to do this, like going with lazy_static for instance.

Comment (epsilon-phase) [reply plume-front/src/editor.rs]: That option looked off to us too tbh.. At least on its own
Maybe it'd be better to use a RefCell or something?

Comment (AnaGelez) [reply plume-front/src/editor.rs]: Anything that will prevent having `unsafe` code is fine (not that it is really unsafe, we only have one thread when running in the browser).

File: plume-front/src/editor.rs (hunk 6 above)
Comment (epsilon-phase) [reply plume-front/src/editor.rs]: Didn't realize it was possible to do that with `i18n!`, that's nifty :)

**Pull request merged**